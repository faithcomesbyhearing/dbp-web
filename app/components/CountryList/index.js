/**
 *
 * CountryList
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { List, AutoSizer } from 'react-virtualized';
import LoadingSpinner from '../LoadingSpinner';

class CountryList extends React.PureComponent {
	constructor(props) {
		super(props);

		this.state = {
			startY: 0,
			distance: 0,
			endY: 0,
			pulling: false,
		};
	}

	getFilteredCountries(width, height) {
		const {
			countries,
			setCountryName,
			setFromCountry,
			toggleLanguageList,
			activeCountryName,
			setCountryListState,
			getCountry,
			filterText,
		} = this.props;
		let countryArray = [];
		if (Array.isArray(countries)) {
			countryArray = countries;
		} else if (countries && typeof countries === 'object') {
			countryArray = Object.values(countries);
		}

		const filteredCountryMap = filterText
			? countryArray.filter((country) =>
					this.filterFunction(country, filterText),
				)
			: countryArray;

		const filteredCountries = filteredCountryMap;

		if (countryArray.length === 0) {
			return (
				<div className={'country-error-message'}>
					There was an error fetching this resource, an Admin has been notified.
					We apologize for the inconvenience.
				</div>
			);
		}

		const renderARow = ({ index, style, key }) => {
			const country = filteredCountries[index];

			return (
				<div
					className="country-name"
					key={key}
					style={style}
					role="button"
					tabIndex={0}
					title={country['name']}
					onClick={() => {
						setFromCountry(true);
						setCountryName({
							name: country['name'],
							languages: country['languages'],
						});
						getCountry({ iso: country?.['codes']?.['iso'] });
						setCountryListState();
						toggleLanguageList();
					}}
				>
					<svg className="icon" height="25px" width="25px">
						<use
							xmlnsXlink="http://www.w3.org/1999/xlink"
							xlinkHref={`/flags.svg#${country?.['codes']?.['iso_a2']}`}
						/>
					</svg>
					<h4
						className={
							activeCountryName === country['name']
								? 'active-language-name'
								: 'inactive-country'
						}
					>
						{country['name']}
					</h4>
				</div>
			);
		};

		const getActiveIndex = () =>
			filteredCountries.findIndex((c) => c.name === activeCountryName)
		;

		return filteredCountries.length ? (
			<List
				id={'list-element'}
				estimatedRowSize={32 * filteredCountries.length}
				height={height}
				rowRenderer={renderARow}
				rowCount={filteredCountries.length}
				overscanRowCount={0}
				rowHeight={32}
				scrollToIndex={getActiveIndex() || 0}
				width={width}
				scrollToAlignment={'start'}
			/>
		) : (
			<div className={'country-error-message'}>
				There are no matches for your search.
			</div>
		);
	}

	// If a swipe happened
	// Check if list's scroll top is 0
	// If it is at 0 then activate the logic for refreshing the countries
	// Otherwise do not do anything
	handleStart = (clientY) => {
		if (this.listScrollTop() === 0) {
			this.setState({ startY: clientY, pulling: true });
		}
	};

	handleMove = (clientY) => {
		// Difference between this move position and the start position
		// is the distance the refresh message should be from where it started
		const maxDistance = 80;
		const minDistance = 0;
		if (
			this.listScrollTop() === 0 &&
			this.state.startY === 0 &&
			!this.state.pulling
		) {
			this.setState({ startY: clientY, pulling: true });
		} else if (
			clientY - this.state.startY <= maxDistance &&
			clientY - this.state.startY >= minDistance &&
			this.state.pulling
		) {
			this.setState((cs) => ({ distance: clientY - cs.startY }));
		}
	};

	handleEnd = (clientY) => {
		// User must have pulled at least 40px
		const minDistance = 40;
		if (
			this.state.startY < clientY &&
			this.state.pulling &&
			this.state.distance > minDistance
		) {
			this.setState({
				startY: 0,
				distance: 0,
				endY: 0,
				pulling: false,
			});
			this.props.getCountries();
		} else {
			this.setState({
				startY: 0,
				distance: 0,
				endY: 0,
				pulling: false,
			});
		}
	};

	handleTouchStart = (touchStartEvent) => {
		this.handleStart(touchStartEvent.targetTouches[0].clientY);
	};

	handleTouchMove = (touchMoveEvent) => {
		this.handleMove(touchMoveEvent.targetTouches[0].clientY);
	};

	handleTouchEnd = (e) => {
		this.handleEnd(e.changedTouches[0].clientY);
	};

	handleMouseDown = (mouseDownEvent) => {
		this.handleStart(mouseDownEvent.clientY);
		this.container.addEventListener('mousemove', this.handleMouseMove);
	};

	handleMouseMove = (mouseMoveEvent) => {
		this.handleMove(mouseMoveEvent.clientY);
	};

	handleMouseUp = (e) => {
		this.handleEnd(e.clientY);
		this.container.removeEventListener('mousemove', this.handleMouseMove);
	};

	handleMouseLeave = (e) => {
		this.handleMouseUp(e);
	};

	listScrollTop = () =>
		document?.getElementById('list-element')
			? document.getElementById('list-element').scrollTop
			: 0;

	filterFunction = (country, filterText) => {
		const lowerCaseText = filterText.toLowerCase();
		return (
			country?.codes?.iso_a2?.toLowerCase().includes(lowerCaseText) ||
			country?.name?.toLowerCase().includes(lowerCaseText)
		);
	};

	handleRef = (el) => {
		this.container = el;
	};

	render() {
		const { active, loadingCountries } = this.props;
		const { distance } = this.state;

		if (active) {
			return (
				<div className="text-selection-section">
					<div
						ref={this.handleRef}
						onTouchStart={this.handleTouchStart}
						onTouchEnd={this.handleTouchEnd}
						onTouchMove={this.handleTouchMove}
						onMouseDown={this.handleMouseDown}
						onMouseUp={this.handleMouseUp}
						onMouseLeave={this.handleMouseLeave}
						className="country-name-list"
					>
						<div
							style={{ height: distance, maxHeight: distance }}
							className={
								distance ? 'pull-down-refresh pulling' : 'pull-down-refresh'
							}
						>
							<span style={{ textAlign: 'center', width: '100%' }}>
								{`${distance > 40 ? 'Release' : 'Pull'} to Refresh`}
							</span>
						</div>
						{loadingCountries ? (
							<LoadingSpinner />
						) : (
							<AutoSizer>
								{({ width, height }) =>
									this.getFilteredCountries(width, height)
								}
							</AutoSizer>
						)}
					</div>
				</div>
			);
		}
		return null;
	}
}

CountryList.propTypes = {
	countries: PropTypes.object,
	setCountryName: PropTypes.func,
	setFromCountry: PropTypes.func,
	toggleLanguageList: PropTypes.func,
	setCountryListState: PropTypes.func,
	getCountry: PropTypes.func,
	getCountries: PropTypes.func,
	filterText: PropTypes.string,
	active: PropTypes.bool,
	loadingCountries: PropTypes.bool,
	activeCountryName: PropTypes.string,
};

export default CountryList;
