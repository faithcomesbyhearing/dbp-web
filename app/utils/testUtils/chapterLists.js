export const textChapters = {
	ENGKJV: {
		filesetType: 'text_plain',
		books: [
			{
				book_id: 'GEN',
				book_id_usfx: 'GN',
				book_id_osis: 'Gen',
				name: 'Genesis',
				testament: 'OT',
				testament_order: 1,
				book_order: 1,
				book_group: 'The Law',
				chapters: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
			},
			{
				book_id: 'EXO',
				book_id_usfx: 'EX',
				book_id_osis: 'Exod',
				name: 'Exodus',
				testament: 'OT',
				testament_order: 2,
				book_order: 2,
				book_group: 'The Law',
				chapters: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
			},
			{
				book_id: 'MAT',
				book_id_usfx: 'MT',
				book_id_osis: 'Matt',
				name: 'Matthew',
				testament: 'NT',
				testament_order: 1,
				book_order: 41,
				book_group: 'Gospels',
				chapters: [11],
			},
			{
				book_id: 'MRK',
				book_id_usfx: 'MK',
				book_id_osis: 'Mark',
				name: 'Mark',
				testament: 'NT',
				testament_order: 2,
				book_order: 42,
				book_group: 'Gospels',
				chapters: [15],
			},
			{
				book_id: 'LUK',
				book_id_usfx: 'LK',
				book_id_osis: 'Luke',
				name: 'Luke',
				testament: 'NT',
				testament_order: 3,
				book_order: 43,
				book_group: 'Gospels',
				chapters: [7, 8, 9, 10],
			},
			{
				book_id: 'JHN',
				book_id_usfx: 'JN',
				book_id_osis: 'John',
				name: 'John',
				testament: 'NT',
				testament_order: 4,
				book_order: 44,
				book_group: 'Gospels',
				chapters: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
			},
		],
	},
};
export const audioChapters = {
	ENGKJVN2DA: {
		filesetType: 'audio_drama',
		books: [
			{
				book_id: 'MAT',
				book_id_usfx: 'MT',
				book_id_osis: 'Matt',
				name: 'Matthew',
				testament: 'NT',
				testament_order: 1,
				book_order: 41,
				book_group: 'Gospels',
				chapters: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
			},
			{
				book_id: 'MRK',
				book_id_usfx: 'MK',
				book_id_osis: 'Mark',
				name: 'Mark',
				testament: 'NT',
				testament_order: 2,
				book_order: 42,
				book_group: 'Gospels',
				chapters: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
			},
			{
				book_id: 'LUK',
				book_id_usfx: 'LK',
				book_id_osis: 'Luke',
				name: 'Luke',
				testament: 'NT',
				testament_order: 3,
				book_order: 43,
				book_group: 'Gospels',
				chapters: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
			},
		],
	},
};
