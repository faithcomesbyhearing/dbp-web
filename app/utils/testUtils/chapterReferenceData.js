export const filesets = [
	{ id: 'ENGESVN2DA', type: 'audio_drama', size: 'NT' },
	{ id: 'ENGESV', type: 'text_plain', size: 'C' },
	{ id: 'ENGESVO2DA', type: 'audio_drama', size: 'OT' },
	{ id: 'ENGESVO1DA', type: 'audio', size: 'OT' },
	{ id: 'ENGESV', type: 'text_format', size: 'C' },
	{ id: 'ENGESVN1DA', type: 'audio', size: 'NT' },
	{ id: 'ENGESVP2DV', type: 'video_stream', size: 'NTP' },
];
export const filesets2 = [
	{ id: 'ENGESVN2DA', type: 'audio_drama', size: 'NT' },
	{ id: 'ENGESV', type: 'text_plain', size: 'C' },
	{ id: 'ENGESVO2DA', type: 'audio_drama', size: 'OT' },
	{ id: 'ENGESVO1DA', type: 'audio', size: 'OT' },
	{ id: 'ENGESV', type: 'text_format', size: 'C' },
	{ id: 'ENGESVN1DA', type: 'audio', size: 'NT' },
];
export const filesets3 = [
	{ id: 'ENGESV', type: 'text_plain', size: 'C' },
	{ id: 'ENGESVO2DA', type: 'audio_drama', size: 'OT' },
	{ id: 'ENGESVO1DA', type: 'audio', size: 'OT' },
	{ id: 'ENGESV', type: 'text_format', size: 'C' },
];
export const filesets4 = [
	{ id: 'ENGESVN2DA', type: 'audio_drama', size: 'NT' },
	{ id: 'ENGESVN1DA', type: 'audio', size: 'NT' },
];
export const filesets5 = [
	{ id: 'ENGESVO2DA', type: 'audio_drama', size: 'OT' },
	{ id: 'ENGESVO1DA', type: 'audio', size: 'OT' },
];
export const filesets6 = [
	{ id: 'ENGESV', type: 'text_plain', size: 'C' },
	{ id: 'ENGESV', type: 'text_format', size: 'C' },
];
export const filesets7 = [
	{ id: 'ENGESV', type: 'text_plain', size: 'C' },
	{ id: 'ENGESV', type: 'text_format', size: 'C' },
];
export const filesets8 = [
	{ id: 'ENGESV', type: 'text_plain', size: 'OT' },
	{ id: 'ENGESV', type: 'text_format', size: 'OT' },
];
export const bookMetaResponse = [
	{
		ENGESVP2DV: {
			filesetType: 'video_stream',
			books: [
				{
					book_id: 'MRK',
					book_id_usfx: 'MR',
					book_id_osis: 'Mrk',
					name: 'Mark',
					testament: 'NT',
					testament_order: 2,
					book_order: 42,
					book_group: 'Gospels',
					chapters: [6, 7],
				},
			],
		},
	},
	{
		ENGESVN2DA: {
			filesetType: 'audio_drama',
			books: [
				{
					book_id: 'MAT',
					book_id_usfx: 'GN',
					book_id_osis: 'Mat',
					name: 'Matthew',
					testament: 'NT',
					testament_order: 1,
					book_order: 41,
					book_group: 'Gospels',
					chapters: [5, 6, 7],
				},
			],
		},
	},
	{
		ENGESV: {
			filesetType: 'text_plain',
			books: [
				{
					book_id: 'GEN',
					book_id_usfx: 'GN',
					book_id_osis: 'Gen',
					name: 'Genesis',
					testament: 'OT',
					testament_order: 1,
					book_order: 1,
					book_group: 'The Law',
					chapters: [1, 2, 3, 4, 5, 6, 7],
				},
				{
					book_id: 'MAT',
					book_id_usfx: 'GN',
					book_id_osis: 'Mat',
					name: 'Matthew',
					testament: 'NT',
					testament_order: 1,
					book_order: 41,
					book_group: 'Gospels',
					chapters: [1, 2, 3, 4, 5, 6, 7],
				},
				{
					book_id: 'PSA',
					book_id_usfx: 'PS',
					book_id_osis: 'Psa',
					name: 'Psalms',
					testament: 'OT',
					testament_order: 1,
					book_order: 29,
					book_group: 'The Law',
					chapters: [70],
				},
			],
		},
	},
	{
		ENGESV: {
			filesetType: 'text_format',
			books: [
				{
					book_id: 'PSA',
					book_id_usfx: 'PS',
					book_id_osis: 'Psa',
					name: 'Psalms',
					testament: 'OT',
					testament_order: 1,
					book_order: 29,
					book_group: 'The Law',
					chapters: [70],
				},
			],
		},
	},
	{
		ENGESVO2DA: {
			filesetType: 'audio_drama',
			books: [
				{
					book_id: 'PSA',
					book_id_usfx: 'PS',
					book_id_osis: 'Psa',
					name: 'Psalms',
					testament: 'OT',
					testament_order: 1,
					book_order: 29,
					book_group: 'The Law',
					chapters: [6, 70],
				},
			],
		},
	},
	{
		ENGESVO1DA: {
			filesetType: 'audio',
			books: [
				{
					book_id: 'GEN',
					book_id_usfx: 'GN',
					book_id_osis: 'Gen',
					name: 'Genesis',
					testament: 'OT',
					testament_order: 1,
					book_order: 1,
					book_group: 'The Law',
					chapters: [1, 2, 3, 4, 5, 6, 7],
				},
			],
		},
	},
	{
		ENGESVN1DA: {
			filesetType: 'audio',
			books: [
				{
					book_id: 'MAT',
					book_id_usfx: 'GN',
					book_id_osis: 'Mat',
					name: 'Matthew',
					testament: 'NT',
					testament_order: 1,
					book_order: 41,
					book_group: 'Gospels',
					chapters: [5, 6, 7],
				},
			],
		},
	},
];
export const bookMetaData = [
	{
		book_id: 'GEN',
		book_id_usfx: 'GN',
		book_id_osis: 'Gen',
		name: 'Genesis',
		testament: 'OT',
		testament_order: 1,
		book_order: 1,
		book_group: 'The Law',
		chapters: [1, 2, 3, 4, 5, 6, 7, 48, 49, 50],
	},
];
