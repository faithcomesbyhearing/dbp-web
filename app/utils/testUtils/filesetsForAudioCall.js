export const engesv = [
	{
		id: 'ENGESVN2DA',
		type: 'audio_drama',
		size: 'NT',
	},
	{
		id: 'ENGESV',
		type: 'text_plain',
		size: 'C',
	},
	{
		id: 'ENGESVC2DA',
		type: 'audio_drama',
		size: 'OT',
	},
	{
		id: 'ENGESVO2DA',
		type: 'audio_drama',
		size: 'OT',
	},
	{
		id: 'ENGESVO1DA',
		type: 'audio',
		size: 'OT',
	},
	{
		id: 'ENGESV',
		type: 'text_format',
		size: 'C',
	},
	{
		id: 'ENGESVN1DA',
		type: 'audio',
		size: 'NT',
	},
];
export const ntpotp = [
	{
		id: 'OREWBTP1DA',
		type: 'audio',
		size: 'OTP',
	},
	{
		id: 'OREWBTS1DA',
		type: 'audio',
		size: 'OTP',
	},
	{
		id: 'OREWBTP2DA',
		type: 'audio_drama',
		size: 'NTP',
	},
];
export const seawbt = [
	{
		id: 'SEAWYIP2DA',
		type: 'audio_drama',
		size: 'NTPOTP',
	},
	{
		id: 'SEAWYIP2DA',
		type: 'app',
		size: 'NTP',
	},
	{
		id: 'SEAWYIP1DA',
		type: 'audio',
		size: 'NTPOTP',
	},
	{
		id: 'SEAWYIP2DA16',
		type: 'audio_drama',
		size: 'NTPOTP',
	},
];
export const filesets3 = [
	{ id: 'ENGESV', type: 'text_plain', size: 'C' },
	{ id: 'ENGESVO2DA', type: 'audio_drama', size: 'OT' },
	{ id: 'ENGESVO1DA', type: 'audio', size: 'OT' },
	{ id: 'ENGESV', type: 'text_format', size: 'C' },
];
export const filesets4 = [
	{ id: 'ENGESVN2DA', type: 'audio_drama', size: 'NT' },
	{ id: 'ENGESVN1DA', type: 'audio', size: 'NT' },
];
export const filesets5 = [
	{ id: 'ENGESVO2DA', type: 'audio_drama', size: 'OT' },
	{ id: 'ENGESVO1DA', type: 'audio', size: 'OT' },
];
export const filesets6 = [
	{ id: 'ENGESV', type: 'text_plain', size: 'C' },
	{ id: 'ENGESV', type: 'text_format', size: 'C' },
];
export const filesets7 = [
	{ id: 'ENGESV', type: 'text_plain', size: 'C' },
	{ id: 'ENGESV', type: 'text_format', size: 'C' },
];
export const filesets8 = [
	{ id: 'ENGESV', type: 'text_plain', size: 'OT' },
	{ id: 'ENGESV', type: 'text_format', size: 'OT' },
];
