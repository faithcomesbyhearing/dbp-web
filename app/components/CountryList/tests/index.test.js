import React from 'react';
import renderer from 'react-test-renderer';
import { fromJS } from 'immutable';
import { countries } from '../../../utils/testUtils/countryData';
import CountryList from '..';

jest.mock('react-virtualized', () => ({
	List: ({ rowRenderer, rowCount }) => {
		const components = [];
		for (let i = 0; i < rowCount; i++) {
			components.push(rowRenderer({ index: i, style: '', key: `${i}_row` }));
		}
		return components;
	},
	AutoSizer: ({ children }) => children({ width: 150, height: 50 }),
}));

const props = {
	countries: fromJS(countries),
	setCountryName: jest.fn(),
	toggleLanguageList: jest.fn(),
	setCountryListState: jest.fn(),
	getCountry: jest.fn(),
	getCountries: jest.fn(),
	filterText: '',
	active: true,
	loadingCountries: false,
	activeCountryName: 'ANY',
};

describe('CountryList component', () => {
	it('should match snapshot of active list', () => {
		const tree = renderer.create(<CountryList {...props} />).toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('should match snapshot of active list with an applied filter', () => {
		const tree = renderer
			.create(<CountryList {...props} filterText={'uni'} />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
