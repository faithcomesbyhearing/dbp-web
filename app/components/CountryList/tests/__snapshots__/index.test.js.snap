// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryList component should match snapshot of active list 1`] = `
<DocumentFragment>
  <div
    class="text-selection-section"
  >
    <div
      class="country-name-list"
    >
      <div
        class="pull-down-refresh"
        style="height: 0px; max-height: 0;"
      >
        <span
          style="text-align: center; width: 100%;"
        >
          Pull to Refresh
        </span>
      </div>
      <div
        class="country-error-message"
      >
        There are no matches for your search.
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CountryList component should match snapshot of active list with an applied filter 1`] = `
<DocumentFragment>
  <div
    class="text-selection-section"
  >
    <div
      class="country-name-list"
    >
      <div
        class="pull-down-refresh"
        style="height: 0px; max-height: 0;"
      >
        <span
          style="text-align: center; width: 100%;"
        >
          Pull to Refresh
        </span>
      </div>
      <div
        class="country-name"
        role="button"
        tabindex="0"
        title="Réunion"
      >
        <svg
          class="icon"
          height="25px"
          width="25px"
        >
          <use
            xlink:href="/flags.svg#RE"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
        </svg>
        <h4
          class="inactive-country"
        >
          Réunion
        </h4>
      </div>
      <div
        class="country-name"
        role="button"
        tabindex="0"
        title="Tunisia"
      >
        <svg
          class="icon"
          height="25px"
          width="25px"
        >
          <use
            xlink:href="/flags.svg#TN"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
        </svg>
        <h4
          class="inactive-country"
        >
          Tunisia
        </h4>
      </div>
      <div
        class="country-name"
        role="button"
        tabindex="0"
        title="United Arab Emirates"
      >
        <svg
          class="icon"
          height="25px"
          width="25px"
        >
          <use
            xlink:href="/flags.svg#AE"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
        </svg>
        <h4
          class="inactive-country"
        >
          United Arab Emirates
        </h4>
      </div>
      <div
        class="country-name"
        role="button"
        tabindex="0"
        title="United Kingdom"
      >
        <svg
          class="icon"
          height="25px"
          width="25px"
        >
          <use
            xlink:href="/flags.svg#GB"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
        </svg>
        <h4
          class="inactive-country"
        >
          United Kingdom
        </h4>
      </div>
      <div
        class="country-name"
        role="button"
        tabindex="0"
        title="United States"
      >
        <svg
          class="icon"
          height="25px"
          width="25px"
        >
          <use
            xlink:href="/flags.svg#US"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
        </svg>
        <h4
          class="inactive-country"
        >
          United States
        </h4>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
